src/message/MessageContainer.js

```diff
@@ -16,29 +18,60 @@ const htmlToDomTree = html => {
   return domTree;
 };
 
-export default class MessageContainer extends React.PureComponent {
+
+class MessageContainer extends React.PureComponent {
+  constructor() {
+    super();
+    this.state = {
+      showActionSheet: false
+    };
+  }
 
   handlePress = () => {
+    const isFromReply = false;
+    this.narrow(isFromReply);
+  };
+
+  narrow = (isFromReply: boolean) => {
     const { message, doNarrow } = this.props;
-    doNarrow(narrowFromMessage(message), message.id);
-  }
+    doNarrow(narrowFromMessage(message), message.id, isFromReply);
+  };
+
+  narrowToConversation = () => {
+    const isFromReply = false;
+    this.narrow(isFromReply);
+  };
+
+  handleLongPress = () => {
+    this.setState({ showActionSheet: true });
+  };
+
+  closeMessageActionSheet = () => {
+    this.setState({ showActionSheet: false });
+  };
 
   render() {
     const { message, auth, avatarUrl, twentyFourHourTime, isBrief, doNarrow } = this.props;
+    const { showActionSheet } = this.state;
     const MessageComponent = isBrief ? MessageBrief : MessageFull;
     const dom = htmlToDomTree(message.content);
 
     return (
-      <MessageComponent
-        message={message}
-        avatarUrl={avatarUrl}
-        twentyFourHourTime={twentyFourHourTime}
-        selfEmail={auth.email}
-        doNarrow={doNarrow}
-        onPress={this.handlePress}
-      >
-        {renderHtmlChildren({ childrenNodes: dom, auth })}
-      </MessageComponent>
+      <View>
+        <MessageComponent
+          message={message}
+          avatarUrl={avatarUrl}
+          twentyFourHourTime={twentyFourHourTime}
+          selfEmail={auth.email}
+          doNarrow={doNarrow}
+          onLongPress={this.handleLongPress}
+        >
+          {renderHtmlChildren({ childrenNodes: dom, auth })}
+        </MessageComponent>
+        <MessageActionSheet show={showActionSheet} close={this.closeMessageActionSheet}/>
```

Comment:
I was imagining MessageActionShet being the child of 'Chat' or some other component. If we keep it here, one of it will exist for every single message.

Maybe the best place for the ActionSheet open/close is in MainScreenContainer.
It will pass an onMessageLongPress handlers to MainScreen, then MainScreen to Chat etc.